crates/radix-engine-toolkit-uniffi/src/builder/transaction_builder.rs

```diff
@@ -77,68 +76,76 @@ impl TransactionBuilderMessageStep {
         self: Arc<Self>,
         message: Message,
     ) -> Arc<TransactionBuilderIntentSignaturesStep> {
-        let header = self.0.clone();
-        let manifest = self.1.clone();
-        Arc::new(TransactionBuilderIntentSignaturesStep(
-            header,
-            manifest,
-            message,
-            vec![],
-            vec![],
-        ))
+        TransactionBuilderIntentSignaturesStep::new(
+            &TransactionBuilderMessageStep(
+                self.0.clone(),
+                self.1.clone(),
+                message,
+            ),
+        )
     }
 
     pub fn sign_with_private_key(
         self: Arc<Self>,
         private_key: Arc<PrivateKey>,
     ) -> Arc<TransactionBuilderIntentSignaturesStep> {
-        let header = self.0.clone();
-        let manifest = self.1.clone();
-        let message = self.2.clone();
-        Arc::new(TransactionBuilderIntentSignaturesStep(
-            header,
-            manifest,
-            message,
-            vec![private_key],
-            vec![],
-        ))
+        let builder = TransactionBuilderIntentSignaturesStep::new(&self);
+        builder.sign_with_private_key(private_key)
     }
 
     pub fn sign_with_signer(
         self: Arc<Self>,
         signer: Box<dyn Signer>,
     ) -> Arc<TransactionBuilderIntentSignaturesStep> {
-        let header = self.0.clone();
-        let manifest = self.1.clone();
-        let message = self.2.clone();
-        Arc::new(TransactionBuilderIntentSignaturesStep(
-            header,
-            manifest,
-            message,
-            vec![],
-            vec![signer],
-        ))
+        let builder = TransactionBuilderIntentSignaturesStep::new(&self);
+        builder.sign_with_signer(signer)
     }
 }
 
 #[uniffi::export]
 impl TransactionBuilderIntentSignaturesStep {
+    #[uniffi::constructor]
+    fn new(message_step: &TransactionBuilderMessageStep) -> Arc<Self> {
+        let intent = Intent {
+            header: message_step.0.clone(),
+            manifest: Arc::new(message_step.1.clone()),
+            message: message_step.2.clone(),
+        };
+        let hash = Hash(intent.hash().unwrap().0);
```

Comment:
@mstrug-rdx and I agreed to split this into a separate PR that removes the `unwrap`s and returns `Result`s instead.